internal/debug/lifecycle.go

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2017-2021 Uber Technologies Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package debug
+
+import "go.uber.org/atomic"
+
+type (
+	// lifeCycleImpl implements the LifeCycle interface
+	lifeCycleImpl struct {
+		pollerCount atomic.Int32
+	}
+
+	// runImpl implements the Run interface
+	runImpl struct {
+		lifeCycle *lifeCycleImpl
+		stopped   atomic.Bool
+	}
```

Comment:
I don't understand why we need this struct. The lifecycle holds the whole state - and where it is call there is no risk of calling stop twice. 

So we could call .Stop() on the LifeCycle itself.